sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,546 @@
+            friend marray operatorOP(marray &rhs) = delete;
+         
+            // OP is: &, |, ^
+            friend marray operatorOP(const marray &lhs, const marray &rhs) = delete;
+            friend marray operatorOP(const marray &lhs, const DataT &rhs) = delete;
+         
+            // OP is: &=, |=, ^=
+            friend marray &operatorOP(marray &lhs, const marray &rhs) = delete;
+            friend marray &operatorOP(marray &lhs, const DataT &rhs) = delete;
+         
+            // OP is: &&, ||
+            friend marray<bool, NumElements> operatorOP(const marray &lhs, const marray &rhs) = delete;
+            friend marray<bool, NumElements> operatorOP(const marray& lhs, const DataT &rhs) = delete;
+         
+            // OP is: <<, >>
+            friend marray operatorOP(const marray &lhs, const marray &rhs) = delete;
+            friend marray operatorOP(const marray &lhs, const DataT &rhs) = delete;
+         
+            // OP is: <<=, >>=
+            friend marray &operatorOP(marray &lhs, const marray &rhs) = delete;
+            friend marray &operatorOP(marray &lhs, const DataT &rhs) = delete;
+         
+            // OP is: ==, !=
+            friend marray<bool, NumElements> operatorOP(const marray &lhs, const marray &rhs) {
+            /* ... */ }
+            friend marray<bool, NumElements> operatorOP(const marray &lhs, const DataT &rhs) {
+            /* ... */ }
+
+            // OP is: <, >, <=, >=
+            friend marray<bool, NumElements> operatorOP(const marray &lhs, const marray &rhs) = delete;
+            friend marray<bool, NumElements> operatorOP(const marray &lhs, const DataT &rhs) = delete;
+         
+            friend marray operator~(const marray &v) = delete;
+         
+            // OP is: +, -, *, /
+            friend marray operatorOP(const DataT &lhs, const marray &rhs) { /* ... */ }
+
+            // OP is: %
+            friend marray operatorOP(const DataT &lhs, const marray &rhs) = delete;
+         
+            // OP is: &, |, ^
+            friend marray operatorOP(const DataT &lhs, const marray &rhs) = delete;
+         
+            // OP is: &&, ||
+            friend marray<bool, NumElements> operatorOP(const DataT &lhs, const marray &rhs) = delete;
+         
+            // OP is: <<, >>
+            friend marray operatorOP(const DataT &lhs, const marray &rhs) = delete;
+         
+            // OP is: ==, !=
+            friend marray<bool, NumElements> operatorOP(const DataT &lhs, const marray &rhs) {
+            /* ... */ }
+
+            // OP is: <, >, <=, >=
+            friend marray<bool, NumElements> operatorOP(const DataT &lhs, const marray &rhs)  = delete;
+         
+            friend marray<bool, NumElements> operator!(const marray &v) = delete;
+    }
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The list of deleted operators are: %, %=, ++, --, +, -, &, |, ^, &=, |=, ^=,
+&&, ||, <<, >>, <<=, >>=, <, >, +<=+, >=, ~, !
```

Comment:
I think there is no need to create a specialization of `marray` for `sycl::ext::oneapi::complex` because the standard SYCL definition should be sufficient.  Note that the standard SYCL `marray` only requires that the underlying type `T` is a C++ [NumericType](https://en.cppreference.com/w/cpp/named_req/NumericType), which is quite expansive.

It looks to me like `sycl::ext::oneapi::complex` does not currently meet the requirements of NumericType because it is missing a default constructor and is missing an assignment operator.  However, these seem like oversights that should be fixed.

You will probably ask about the `marray` operators that make no sense for complex types.  This, I think, is a bug in the SYCL spec that should be fixed.  Essentially, the spec should say that each operator is only available if the underlying type `T` implements that operator.  See https://gitlab.khronos.org/sycl/Specification/-/issues/559 (requires Khronos access).

Reply:
Okay, I have updated this proposal to use the different `marray` definition. Instead this proposal just instantiates the `marray` class for the various `gencomplex` types.